### PR TITLE
RMC-173 Blank Questionnaires Revenge

### PR DESCRIPTION
**Pull request by LeoKHoward**
# Motivation and Context
Need to identify when a blank questionnaire is returned so the appropriate action can be taken as per the BQ rules table in Confluence

# What has changed
New feature created specifically for blank questionnaires
Two primary scenarios created - standard blank questionnaire receipting and blank questionnaires before actual receipting

# How to test?
Clone branch repos for card (case-processor/fieldwork-adapter/ddl/pubsub), build the images and run the ATs

# Links
https://trello.com/c/EmrziSj3/439-rmc-173-blank-questionnaire-returned-21
https://collaborate2.ons.gov.uk/confluence/display/SDC/Process+Flow+for+Blank+PQs
https://github.com/ONSdigital/census-rm-fieldwork-adapter/pull/38
https://github.com/ONSdigital/census-rm-case-processor/pull/128
https://github.com/ONSdigital/census-rm-ddl/pull/47
https://github.com/ONSdigital/census-rm-pubsub/pull/40
https://github.com/ONSdigital/census-rm-acceptance-tests/pull/222

### CODE REVIEW

**Comment by LeoKHoward** [comment]
[RMC-173 Blank Questionnaire Returned (21)](https://trello.com/c/EmrziSj3/439-rmc-173-blank-questionnaire-returned-21)

**Review by NickGrantX — APPROVED** [review]
ATs passing. Great work!

File: acceptance_tests/features/steps/receipt.py [hunk 1]
```diff
@@ -244,12 +284,14 @@ def check_ce_actual_responses_and_receipted(context, incremented, receipted, cas
 
 @step('if the field instruction "{action_instruction_type}" is not NONE a msg to field is emitted'
       ' where ceActualResponse is incremented "{incremented}"')
-def check_receipt_to_field_msg(context, action_instruction_type, incremented):
+@step('if the field instruction "{action_instruction_type}" is not NONE a msg to field is emitted')
+@step('the field instruction is "{action_instruction_type}"')
+def check_receipt_to_field_msg(context, action_instruction_type, incremented=None):
```
Comment (liamtoozer) [inline acceptance_tests/features/steps/receipt.py:303]: Think `incremented` isn't used here?

File: acceptance_tests/features/steps/receipt.py [hunk 2]
```diff
@@ -207,7 +238,16 @@ def get_new_qid_and_case_as_required(context, case_type, address_level, qid_type
...
 
 
-@when("the receipt msg is put on the GCP pubsub")
+@step('if required for "{questionnaire_type}", a new qid is created "{qid_needed}"')
+def get_second_qid(context, questionnaire_type, qid_needed):
+    if qid_needed == 'True':
+        generate_post_request_body(context, questionnaire_type)
+        listen_for_ad_hoc_uac_updated_message(context, questionnaire_type)
+        context.qid_to_receipt = context.requested_qid
+        return
```
Comment (liamtoozer) [inline acceptance_tests/features/steps/receipt.py]: `return` could prob be removed here?

File: acceptance_tests/features/steps/receipt.py (hunk 1 above)
Comment (LeoKHoward) [reply acceptance_tests/features/steps/receipt.py:303]: It is used by the top step and taking it out causes most of scenarios in receipt.feature to fail

File: acceptance_tests/features/steps/receipt.py (hunk 2 above)
Comment (LeoKHoward) [reply acceptance_tests/features/steps/receipt.py]: Removed

File: acceptance_tests/features/steps/receipt.py (hunk 1 above)
Comment (liamtoozer) [reply acceptance_tests/features/steps/receipt.py:303]: Looks like it's not being used at all by [this last step](https://github.com/ONSdigital/census-rm-acceptance-tests/blob/255d1151f4232ac7b678bee440942a8e8116062a/acceptance_tests/features/receipt.feature#L10), so can be removed from there as well. I think this was to check receipt count if it was an individual case (but looks like it's not needed for this step)

Comment (liamtoozer) [reply acceptance_tests/features/steps/receipt.py:303]: This wasn't really to do with this PR though, so can probably be tidied up in future when someone is in the receipt feature next 👍

File: acceptance_tests/features/blank_questionnaire.feature [hunk 3]
```diff
@@ -0,0 +1,117 @@
+Feature: Handling Blank Questionnaire Scenario
+
+  Scenario Outline: Blank questionnaire for non-CE case types
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the offline receipt msg for the receipted case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
+    And a case_updated msg of type "<case type>" and address level "<address level>" is emitted where "receiptReceived" is "True" and qid is "<qid needed>"
+    And if the field instruction "<offline receipt instruction>" is not NONE a msg to field is emitted
+    And if required for "<form type>", a new qid is created "<qid needed>"
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature]: I had trouble working out what was going on with `qid needed`, I think what it's doing is linking another QID to the case and receipting it to test that scenario, maybe a name like `other qid receipted` would help

File: acceptance_tests/features/blank_questionnaire.feature [hunk 4]
```diff
@@ -0,0 +1,117 @@
+Feature: Handling Blank Questionnaire Scenario
+
+  Scenario Outline: Blank questionnaire for non-CE case types
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the offline receipt msg for the receipted case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
+    And a case_updated msg of type "<case type>" and address level "<address level>" is emitted where "receiptReceived" is "True" and qid is "<qid needed>"
+    And if the field instruction "<offline receipt instruction>" is not NONE a msg to field is emitted
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature]: Given how long these scenarios have ended up, I think it's worth removing the steps like this that aren't explicitly relevant to the blank questionnaire feature?

File: acceptance_tests/features/blank_questionnaire.feature [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      | HI        | U             | Ind      | 21        | sample_1_english_HH_unit.csv  | SAMPLE_LOADED,FULFILMENT_REQUESTED                               | NONE              | False      | E       | NONE                        | RM_UAC_CREATED,RESPONSE_RECEIVED,RESPONSE_RECEIVED |
+
+
+  Scenario Outline: Blank questionnaire for CE case types
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the offline receipt msg for the receipted case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
+    And a case_updated msg of type "<case type>" and address level "<address level>" is emitted where "receiptReceived" is "True" and qid is "<qid needed>"
+    And if the field instruction "<offline receipt instruction>" is not NONE a msg to field is emitted
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature]: Given how long these scenarios have ended up, I think it's worth removing the steps like this that aren't explicitly relevant to the blank questionnaire feature

File: acceptance_tests/features/blank_questionnaire.feature [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      | CE        | E             | Ind      | 21        | sample_1_english_CE_estab.csv | SAMPLE_LOADED,RM_UAC_CREATED,FULFILMENT_REQUESTED,RESPONSE_RECEIVED,RESPONSE_RECEIVED                | NONE              | False      | E       | UPDATE                      |
+
+
+  Scenario Outline: Blank questionnaire before actual receipt
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the blank questionnaire msg for a case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
+    And a case_updated msg of type "<case type>" and address level "<address level>" is emitted where "receiptReceived" is "False" and qid is "<qid needed>"
+    And if the field instruction "<instruction>" is not NONE a msg to field is emitted
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature]: Given how long these scenarios have ended up, I think it's worth removing the steps like this that aren't explicitly relevant to the blank questionnaire feature

File: acceptance_tests/features/blank_questionnaire.feature [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      | HH        | U             | HH       | 01        | sample_1_english_HH_unit.csv  | SAMPLE_LOADED,RESPONSE_RECEIVED,RESPONSE_RECEIVED | False      | E       | UPDATE      |                                                    |
+      | SPG       | U             | HH       | 01        | sample_1_english_SPG_unit.csv | SAMPLE_LOADED,RESPONSE_RECEIVED,RESPONSE_RECEIVED | False      | E       | UPDATE      |                                                    |
+      | HI        | U             | Ind      | 21        | sample_1_english_HH_unit.csv  | SAMPLE_LOADED,FULFILMENT_REQUESTED                | False      | E       | NONE        | RM_UAC_CREATED,RESPONSE_RECEIVED,RESPONSE_RECEIVED |
+
+
+  Scenario Outline: Blank questionnaire for Individual qid types before actual receipt
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the blank questionnaire msg for a case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature:71]: Given how long these scenarios have ended up, I think it's worth removing the steps like this that aren't explicitly relevant to the blank questionnaire feature

File: acceptance_tests/features/blank_questionnaire.feature [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      | CE        | E             | Ind      | 21        | sample_1_english_CE_estab.csv | SAMPLE_LOADED,RM_UAC_CREATED,FULFILMENT_REQUESTED,RESPONSE_RECEIVED,RESPONSE_RECEIVED | False      | E       | NONE        |
+      | HH        | U             | HH       | 01        | sample_1_english_HH_unit.csv  | SAMPLE_LOADED,RM_UAC_CREATED,RESPONSE_RECEIVED,RESPONSE_RECEIVED                      | True       | E       | UPDATE      |
+
+  Scenario Outline: Blank questionnaire for non-CE case types before actual receipt when another qid is needed
+    Given sample file "<sample file>" is loaded successfully
+    And if required, a new qid and case are created for case type "<case type>" address level "<address level>" qid type "<qid type>" and country "<country>"
+    And the blank questionnaire msg for a case is put on the GCP pubsub
+    And a uac_updated msg is emitted with active set to false for the receipted qid
+    And a case_updated msg of type "<case type>" and address level "<address level>" is emitted where "receiptReceived" is "False" and qid is "<qid needed>"
+    And if the field instruction "<instruction>" is not NONE a msg to field is emitted
```
Comment (AdamHawtin) [inline acceptance_tests/features/blank_questionnaire.feature]: Given how long these scenarios have ended up, I think it's worth removing the steps like this that aren't explicitly relevant to the blank questionnaire feature

**Review by AdamHawtin — COMMENTED** [review]
Highlighted some steps I think could be removed, though if they make the whole test easier to right by grabbing data for the context etc. then I suppose they are justified

**Review by liamtoozer — APPROVED** [review]
All passes for me - nice!

File: acceptance_tests/features/blank_questionnaire.feature (hunk 4 above)
Comment (ryangrundy7) [reply acceptance_tests/features/blank_questionnaire.feature]: Think it's going to be a lot of hassle and code change to edit these. some context is needed for the steps afterwards

File: acceptance_tests/features/blank_questionnaire.feature (hunk 5 above)
Comment (ryangrundy7) [reply acceptance_tests/features/blank_questionnaire.feature]: Think it's going to be a lot of hassle and code change to edit these. some context is needed for the steps afterwards

File: acceptance_tests/features/blank_questionnaire.feature (hunk 6 above)
Comment (ryangrundy7) [reply acceptance_tests/features/blank_questionnaire.feature]: Think it's going to be a lot of hassle and code change to edit these. some context is needed for the steps afterwards

File: acceptance_tests/features/blank_questionnaire.feature (hunk 7 above)
Comment (ryangrundy7) [reply acceptance_tests/features/blank_questionnaire.feature:71]: Think it's going to be a lot of hassle and code change to edit these. some context is needed for the steps afterwards

File: acceptance_tests/features/blank_questionnaire.feature (hunk 8 above)
Comment (ryangrundy7) [reply acceptance_tests/features/blank_questionnaire.feature]: Think it's going to be a lot of hassle and code change to edit these. some context is needed for the steps afterwards

**Pull request merged**